src/drivers/wa-sqlite/adapter.ts

```diff
@@ -0,0 +1,180 @@
+import { Mutex } from 'async-mutex'
+
+export class DatabaseAdapter implements DatabaseAdapterInterface {
+  db: Database
+  txMutex: Mutex
+
+  constructor(db: Database) {
+    this.db = db
+    this.txMutex = new Mutex()
+  }
+
+  async runInTransaction(...statements: Statement[]): Promise<RunResult> {
+    // Uses a mutex to ensure that transactions are not interleaved.
+    // This is needed because transactions are executed manually using `BEGIN` and `COMMIT` commands.
+    const release = await this.txMutex.acquire()
+    let open = false
+    let rowsAffected = 0
+    try {
+      await this.db.exec({ sql: 'BEGIN' })
+      open = true
+      for (const stmt of statements) {
+        await this.db.exec(stmt)
+        if (isInsertUpdateOrDeleteStatement(stmt.sql)) {
+          // Fetch the number of rows affected by the last insert, update, or delete
+          rowsAffected += await this.db.getRowsModified()
+        }
+      }
+      return {
+        rowsAffected: rowsAffected,
+      }
+    } catch (error) {
+      await this.db.exec({ sql: 'ROLLBACK' })
+      open = false
+      throw error // rejects the promise with the reason for the rollback
+    } finally {
+      if (open) {
+        await this.db.exec({ sql: 'COMMIT' })
+      }
+      release()
+    }
+  }
+
+  async transaction<T>(
+    f: (_tx: Tx, setResult: (res: T) => void) => void
+  ): Promise<T | void> {
+    const release = await this.txMutex.acquire()
+
+    try {
+      await this.db.exec({ sql: 'BEGIN' })
+    } catch (e) {
+      release()
+      throw e
+    }
+
+    return new Promise((resolve, reject) => {
+      const releaseAndReject = (err?: any) => {
+        // if the tx is rolled back, release the lock and reject the promise
+        release()
+        reject(err)
+      }
+
+      const tx = new Transaction(this, releaseAndReject)
+
+      f(tx, (res) => {
+        // Commit the transaction when the user sets the result.
+        // This assumes that the user does not execute any more queries after setting the result.
+        this.db
+          .exec({ sql: 'COMMIT' })
+          .then(() => {
+            release()
+            resolve(res)
+          })
+          .catch((err) => {
+            release()
+            reject(err)
+          })
+      })
+    })
+  }
+
+  run(stmt: Statement): Promise<RunResult> {
+    // Also uses the mutex to avoid running this query while a transaction is executing.
+    // Because that would make the query part of the transaction which was not the intention.
+    return this.txMutex.runExclusive(() => {
+      return this._runUncoordinated(stmt)
+    })
+  }
+
+  // Do not use this uncoordinated version directly!
+  // It is only meant to be used within transactions.
+  async _runUncoordinated(stmt: Statement): Promise<RunResult> {
+    await this.db.exec(stmt)
+    return {
+      rowsAffected: this.db.getRowsModified(),
+    }
+  }
+
+  query(stmt: Statement): Promise<Row[]> {
```

Comment:
Currently, `query()` is to be called whenever you need the result of the query but nothing prevents you from writing a query that modifies something and then reads it. For example, `INSERT INTO ...; SELECT ...`. 

It is debatable whether we want to accept this or not. One could argue that `query` should only be used for single read-only SQL statements and that the previous example of inserting and then reading should use an interactive transaction. But this seems to be an unnecessary restriction.

Btw, `BEGIN READ ONLY` does not seem to exist in SQLite: [https://www.sqlite.org/lang_transaction.html](https://www.sqlite.org/lang_transaction.html)

What do you mean by shared lock? All operations (`runInTransaction`, `transaction`, `run`, and `query` share the same `txMutex`.

Reply:
> Currently, query() is to be called whenever you need the result of the query but nothing prevents you from writing a query that modifies something and then reads it.

I think other drivers enforce that query is for read-only transactions (I would need to check if this is indeed true). In that case, we must offer an uniform interface and document the behaviour. 

> What do you mean by shared lock? 
We could use a shared (as in multiple readers, single writer) lock to allow read-only transactions in parallel.

Reply:
> I think other drivers enforce that query is for read-only transactions (I would need to check if this is indeed true). In that case, we must offer an uniform interface and document the behaviour.

The adapter for cordova, expo, and react-native indeed enforce this by executing the query inside a read-only transaction (provided by those drivers). The adapters for better-sqlite3 and absurd-sql do not seem to enforce it as they just (prepare and) execute the query. I can take a look to make sure all drivers enforce it through `sqlite3_stmt_readonly` although it won't be perfect, see: [https://www.sqlite.org/c3ref/stmt_readonly.html](https://www.sqlite.org/c3ref/stmt_readonly.html).

Now, I'm wondering how cordova, expo, and react-native enforce that transactions are read-only.
If they use `sqlite3_stmt_readonly` it is also not 100% reliable.
Perhaps a more robust way to enforce it would be to have 2 DB connections, one that is read/write and one that is read-only. And execute `query()` using the read-only connection.

> We could use a shared (as in multiple readers, single writer) lock to allow read-only transactions in parallel.
I don't think this is safe. I we don't lock on reads, that means a read may be executed in the middle of a concurrent transaction (because `db.exec` will just think the query is part of the transaction as it is executed after `BEGIN` and before `COMMIT`). Hence, it may read some tentative state while it should only be allowed to read the state before the transaction or the state after the concurrent transaction committed, but nothing in-between.

Reply:
Will do.

Take aways from our conversation over discord:
- need for locking mainly comes from the fact that the front-end and the satellite process are sharing the same DB connection and SQLite does not provide isolation between transactions on the same DB connection
- Need to properly solve this problem by using different DB connections for front-end and satellite process
- Afterwards, we may optimize the drivers to use a shared-lock